server/src/main/java/org/elasticsearch/search/aggregations/bucket/filter/QueryToFilterAdapter.java

```diff
@@ -0,0 +1,343 @@
+import org.apache.lucene.search.Weight;
+import org.apache.lucene.util.Bits;
+import org.elasticsearch.common.CheckedSupplier;
+import org.elasticsearch.common.lucene.Lucene;
+
+import java.io.IOException;
+import java.util.function.BiConsumer;
+import java.util.function.IntPredicate;
+
+/**
+ * Adapts a Lucene {@link Query} to the behaviors used be the
+ * {@link FiltersAggregator}. In general we try to delegate to {@linkplain Query}
+ * when we don't have 
+ */
+public abstract class QueryToFilterAdapter {
+    /**
+     * Build a filter against the provided searcher.
+     * <p>
+     * Note: This method rewrites the query against the {@link IndexSearcher}
+     */
+    public static QueryToFilterAdapter build(IndexSearcher searcher, String key, Query query) throws IOException {
+        query = searcher.rewrite(query);
+        if (query instanceof TermQuery) {
+            return new TermFilter(searcher, key, (TermQuery) query);
+        }
+        if (query instanceof MatchAllDocsQuery) {
+            return new MatchAllQueryToFilterAdapter(searcher, key);
+        }
+        if (query instanceof MatchNoDocsQuery) {
+            return new MatchNoneQueryToFilterAdapter(searcher, key);
+        }
+        return new DefaultQueryToFilterAdapter(searcher, key, query);
+    }
+
+    private final IndexSearcher searcher;
+    private final String key;
+
+    private QueryToFilterAdapter(IndexSearcher searcher, String key) {
+        this.searcher = searcher;
+        this.key = key;
+    }
+
+    public final String key() {
+        return key;
+    }
+
+    protected final IndexSearcher searcher() {
+        return searcher;
+    }
+
+    /**
+     * Is it safe to use index metadata like
+     * {@link IndexReader#docFreq} or {@link IndexReader#maxDoc} to count. the
+     * number of matching documents.
+     */
+    protected final boolean countCanUseMetadata(FiltersAggregator.Counter counter, Bits live) {
+        return live == null && counter.docCount.alwaysOne();
+    }
+
+    /**
+     * Make a filter that matches this filter and the provided query.
+     * <p>
+     * Note: This method rewrites the query against the {@link IndexSearcher}.
+     */
+    abstract QueryToFilterAdapter union(Query extraQuery) throws IOException;
+
+    abstract IntPredicate matchingDocIds(LeafReaderContext ctx) throws IOException;
+
+    abstract long count(LeafReaderContext ctx, FiltersAggregator.Counter counter, Bits live) throws IOException;
+
+    /**
+     * Estimate the cost of calling {@code #count} on this leaf.
+     */
+    abstract long estimateCountCost(LeafReaderContext ctx, CheckedSupplier<Boolean, IOException> canUseMetadata) throws IOException;
+
+    /**
+     * Collect profiling information for this filter.
+     */
+    void collectDebugInfo(BiConsumer<String, Object> add) {}
+
+    /**
+     * Special case when the filter can't match anything.
+     */
+    private static class MatchNoneQueryToFilterAdapter extends QueryToFilterAdapter {
```

Comment:
We don't strictly need this for the PR but it feels like a nice example of how you'd implement this for a filter that can come purely from metadata.